### PR TITLE
add controller to autoregister APIServices

### PR DESCRIPTION
**Pull request by deads2k**
As we look to combine the aggregator and the core kube-apiserver which includes things like TPRs, we need a way to reliably auto-manage APIServices for the aggregator so that discovery works as expected and so that users don't have to maintain particular resources that they shouldn't have to.

This controller deals with a dynamic set of APIServices which are synchronized.  The mutation of that set is exposed via threadsafe methods.  If a user *wants* to take control, he can remove our label and do whatever he wants.

The controller approach provides the functionality we need while preserving the independence of various layers.

@kubernetes/sig-api-machinery-pr-reviews @ncdc

### CODE REVIEW

**Comment by deads2k** [comment]
@k8s-bot cvm gce e2e test this

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+)
+
+var (
+	cloner = conversion.NewCloner()
+)
+
+// AutoAPIServiceRegistration is an interface which callers can re-declare locally and properly cast to for
+// adding and removing APIServices
+type AutoAPIServiceRegistration interface {
+	AddAPIServiceToSync(in *apiregistration.APIService)
```
Comment (ncdc) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: godoc on these 2 definitions

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	RemoveAPIServiceToSync(name string)
+}
+
+// AutoRegisterController is used to keep a particular set of APIServices present in the API.  It is useful
+// for cases where you want to auto-register APIs like TPRs or core API server
```
Comment (ncdc) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: "or core API server" ... ?

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+type AutoRegisterController struct {
+	apiServiceLister listers.APIServiceLister
+	apiServiceSynced cache.InformerSynced
+	apiServiceClient apiregistrationclient.APIServicesGetter
+
+	apiServiceToSyncLock sync.RWMutex
```
Comment (ncdc) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: Pluralize services? `apiServicesToSyncLock`

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+	apiServicesToSync    map[string]*apiregistration.APIService
+
+	syncHandler func(apiServiceName string) error
+
+	// queue is where incoming work is placed to de-dup and to allow "easy" rate limited requeues on errors
+	queue workqueue.RateLimitingInterface
+}
+
+func NewAutoRegisterAutoRegisterController(apiServiceInformer informers.APIServiceInformer, apiServiceClient apiregistrationclient.APIServicesGetter) *AutoRegisterController {
```
Comment (ncdc) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: Did you mean to have AutoRegister duplicated in the function name?

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+// AutoAPIServiceRegistration is an interface which callers can re-declare locally and properly cast to for
+// adding and removing APIServices
+type AutoAPIServiceRegistration interface {
+	AddAPIServiceToSync(in *apiregistration.APIService)
+	RemoveAPIServiceToSync(name string)
+}
+
+// AutoRegisterController is used to keep a particular set of APIServices present in the API.  It is useful
+// for cases where you want to auto-register APIs like TPRs or core API server
+type AutoRegisterController struct {
```
Comment (ncdc) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: Do you need this type exported?

Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: > Do you need this type exported?

I can make it private to start.  I suspect that most callers will use the interface and call the run method.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go (hunk 1 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: doing

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go (hunk 2 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: > "or core API server" ... ?

clarifying.  "or groups from the core kube-apiserver"

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go (hunk 3 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: sure

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go (hunk 4 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: fixing

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go (hunk 5 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: doing

**Comment by deads2k** [comment]
comments so far addressed.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+			cast, ok := obj.(*apiregistration.APIService)
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					glog.V(2).Infof("Couldn't get object from tombstone %#v", obj)
+					return
+				}
+				cast, ok = tombstone.Obj.(*apiregistration.APIService)
+				if !ok {
+					glog.V(2).Infof("Tombstone contained object that is not a CSR: %#v", obj)
```
Comment (ncdc) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: CSR? 😄

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+
+	// we aren't trying to manage this APIService.  If the user removes the label, he's taken over management himself
+	case curr.Labels[AutoRegisterManagedLabel] != "true":
+		return nil
+
+	// we have a spurious APIService that we're managing, delete it
+	case desired == nil:
+		return c.apiServiceClient.APIServices().Delete(curr.Name, nil)
+
+		// if the spec's already match, nothing for us to do
```
Comment (ncdc) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: s/spec's/specs/

indentation off

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+	case reflect.DeepEqual(curr.Spec, desired.Spec):
```
Comment (ncdc) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: is reflect ok or do we need semantic?

Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: > is reflect ok or do we need semantic?

Semantic isn't cleanly divorced from a global scheme yet. Reflect works for now.

Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go]: > is reflect ok or do we need semantic?

Semantic isn't cleanly divorced from a global scheme yet. Reflect works for now.

**Comment by deads2k** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by deads2k** [comment]
@ncdc did you make it through?

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller_test.go [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+			},
+		},
+		{
+			name: "modifying it should result in stomping",
+			steps: []func(c AutoAPIServiceRegistration, fakeWatch *watch.FakeWatcher, client internalclientset.Interface){
+				// adding one to auto-register should create
+				func(c AutoAPIServiceRegistration, fakeWatch *watch.FakeWatcher, client internalclientset.Interface) {
+					c.AddAPIServiceToSync(&apiregistration.APIService{ObjectMeta: metav1.ObjectMeta{Name: "foo"}})
+				},
+				// adding an API service which claims to be managed but isn't should be deleted
```
Comment (ncdc) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller_test.go]: fix comment, then lgtm

**Comment by ncdc** [comment]
LGTM

**Comment by deads2k** [comment]
@k8s-bot gce etcd3 e2e test this

**Pull request closed without merging**